### PR TITLE
Rapid7 Nexpose bug fix

### PR DESCRIPTION
**Pull request by Bargenish**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/31019
fixes: https://github.com/demisto/etc/issues/32849

## Description
- Added the *download_immediately* argument to the listed commands to support reports which take longer than 10 seconds to be generated:
    - ***nexpose-create-sites-report***
    - ***nexpose-create-scan-report***
    - ***nexpose-create-assets-report***
- Added the ***nexpose-get-report-status*** which returns the status of a report generation process.
- Added the ***nexpose-download-report*** which downloads a generated report.
- Added the 
Fixed the TPB and removed the scan tasks due to license limitations.
COnverted the README to markdown

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [X] No

## Must have
- [X] Tests
- [X] Documentation

### CODE REVIEW

File: Packs/Rapid7_Nexpose/Playbooks/playbook-Nexpose_-_GenericPolling_-_Create_and_download_a_report.yml [hunk 1]
```diff
@@ -0,0 +1,436 @@
+id: Nexpose - Create and download a report
```
Comment (yaakovi) [inline Packs/Rapid7_Nexpose/Playbooks/playbook-Nexpose_-_GenericPolling_-_Create_and_download_a_report.yml:1]: see [naming conventions](https://xsoar.pan.dev/docs/playbooks/playbook-conventions#playbooks)
update the file name to match the playbook name
```suggestion
id: Nexpose - Create and Download Report
```

File: Packs/Rapid7_Nexpose/Playbooks/playbook-Nexpose_-_GenericPolling_-_Create_and_download_a_report.yml [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+      script: '|||nexpose-download-report'
+      type: regular
+      iscommand: true
+      brand: ""
+    nexttasks:
+      '#none#':
+      - "6"
+    scriptarguments:
+      format:
+        simple: ${Nexpose.Report.Format}
```
Comment (yaakovi) [inline Packs/Rapid7_Nexpose/Playbooks/playbook-Nexpose_-_GenericPolling_-_Create_and_download_a_report.yml:222]: switch to a complex format

File: Packs/Rapid7_Nexpose/Playbooks/playbook-Nexpose_-_GenericPolling_-_Create_and_download_a_report.yml [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+      iscommand: true
+      brand: ""
+    nexttasks:
+      '#none#':
+      - "5"
+    scriptarguments:
+      download_immediately:
+        simple: "false"
+      format:
+        simple: ${inputs.format}
```
Comment (yaakovi) [inline Packs/Rapid7_Nexpose/Playbooks/playbook-Nexpose_-_GenericPolling_-_Create_and_download_a_report.yml:263]: same

File: Packs/Rapid7_Nexpose/Playbooks/playbook-Nexpose_-_GenericPolling_-_Create_and_download_a_report.yml [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+      script: '|||nexpose-create-assets-report'
+      type: regular
+      iscommand: true
+      brand: ""
+    nexttasks:
+      '#none#':
+      - "5"
+    scriptarguments:
+      assets:
+        simple: ${inputs.assets}
```
Comment (yaakovi) [inline Packs/Rapid7_Nexpose/Playbooks/playbook-Nexpose_-_GenericPolling_-_Create_and_download_a_report.yml:302]: same

File: Packs/Rapid7_Nexpose/Playbooks/playbook-Nexpose_-_GenericPolling_-_Create_and_download_a_report.yml [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+      iscommand: true
+      brand: ""
+    nexttasks:
+      '#none#':
+      - "5"
+    scriptarguments:
+      download_immediately:
+        simple: "false"
+      format:
+        simple: ${inputs.format}
```
Comment (yaakovi) [inline Packs/Rapid7_Nexpose/Playbooks/playbook-Nexpose_-_GenericPolling_-_Create_and_download_a_report.yml:344]: same

File: Packs/Rapid7_Nexpose/Playbooks/playbook-Nexpose_-_GenericPolling_-_Create_and_download_a_report.yml [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+  playbookInputQuery: null
+- key: scan
+  value: {}
+  required: false
+  description: integer <int64> The identifier of the scan.
+  playbookInputQuery: null
+- key: name
+  value: {}
+  required: false
+  description: ' The report name.'
```
Comment (yaakovi) [inline Packs/Rapid7_Nexpose/Playbooks/playbook-Nexpose_-_GenericPolling_-_Create_and_download_a_report.yml:402]:
```suggestion
  description: The report name.
```

File: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml [hunk 7]
```diff
@@ -847,7 +904,7 @@ script:
       type: number
     - contextPath: Nexpose.Scan.Vulnerabilities.Total
       description: The total number of vulnerabilities.
-      type: number
+      type: numberFF
```
Comment (yaakovi) [inline Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml:907]:
```suggestion
      type: number
```

File: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml [hunk 8]
```diff
@@ -711,6 +749,13 @@ script:
       - text
       required: false
       secret: false
+    - auto: PREDEFINED
+      defaultValue: true
```
Comment (yaakovi) [inline Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml:753]: for all
```suggestion
      defaultValue: 'true'
```

File: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml [hunk 9]
```diff
@@ -932,6 +989,68 @@ script:
...
+    - name: name
+      description: The report name.
+    - name: format
+      auto: PREDEFINED
+      predefined:
+      - pdf
+      - rtf
+      - xml
+      - html
+      - text
```
Comment (yaakovi) [inline Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml:1010]: switch to upper case

File: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml [hunk 10]
```diff
@@ -932,6 +989,68 @@ script:
...
+    - contextPath: InfoFile.Info
+      description: Info about the report file
+      type: string
+    - contextPath: InfoFile.Size
+      description: Size of the report file
+      type: number
+    - contextPath: InfoFile.Type
+      description: Type of the report file
+      type: string
+    description: Returns the contents of a generated report.
```
Comment (yaakovi) [inline Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml:1032]: the output is a file, not the content of it.
```suggestion
    description: Returns the generated report.
```

File: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.py [hunk 11]
```diff
@@ -1040,59 +1039,127 @@ def create_assets_report_command():
     assets = str(demisto.args()['assets']).split(',')
     template = demisto.args().get('template')
     name = demisto.args().get('name', 'report ' + str(datetime.now()))
-    report_format = demisto.args().get('format', 'pdf')
+    report_format = demisto.args().get('format') or 'pdf'
```
Comment (yaakovi) [inline Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.py:1042]: why did you change it?

File: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.py [hunk 12]
```diff
@@ -1040,59 +1039,127 @@ def create_assets_report_command():
...
+        outputs_key_field=['ID', 'InstanceID'],
+    )
 
 
 def create_report(scope, name, template, report_format):
-    if template is None:
+    if not template:
         templates = get_report_templates()
-        if (templates is None or len(templates) == 0 or 'resources' not in templates):
+        if templates is None or len(templates) == 0 or 'resources' not in templates:
```
Comment (yaakovi) [inline Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.py:1162]:
```suggestion
        if not templates or 'resources' not in templates:
```

File: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.py [hunk 13]
```diff
@@ -1110,31 +1177,58 @@ def create_report(scope, name, template, report_format):
...
 
 
-def download_report(report_id, name, report_format):
-    # Generate the report
-    path = 'reports/' + str(report_id) + ' /generate'
+def generate_report(report_id):
+    path = 'reports/{}/generate'.format(report_id)
     instance = send_request(path, 'post')
-    if (instance is None):
+    if not instance.get('id'):
```
Comment (yaakovi) [inline Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.py:1186]: can `instance` be None here? if so, the `get` will raise an `AttributeError`

File: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.py [hunk 14]
```diff
@@ -1322,11 +1416,15 @@ def main():
         if demisto.command() == 'nexpose-get-report-templates':
             demisto.results(get_report_templates_command())
```
Comment (yaakovi) [inline Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.py:1417]: you can replace everything with `return_results`, it will act the same.

**Review by yaakovi — CHANGES_REQUESTED** [review]

File: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.py (hunk 13 above)
Comment (Bargenish) [reply Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.py:1186]: No, if it's None then the request should fail.
Should I add a verification condition before?

File: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.py (hunk 11 above)
Comment (Bargenish) [reply Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.py:1042]: When executing it through the playbook the 'format' key exists so the default value doesn't get picked.

File: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml (hunk 9 above)
Comment (Bargenish) [reply Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml:1010]: The original commands use these values and I wanted to be consistent. Should I change it anyway?

Comment (yaakovi) [reply Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml:1010]: leave it as is

**Comment by yaakovi** [comment]
@kirbles19 please go over:
* MD files: 3 READMEs and. release notes
* playbook file: Packs/Rapid7_Nexpose/Playbooks/playbook-Nexpose_-_GenericPolling_-_Create_and_Download_Report.yml
* integration yml: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml

**Review by yaakovi — COMMENTED** [review]
looks good, let's do a short demo for the playbook

**Review by yaakovi — APPROVED** [review]

File: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml [hunk 15]
```diff
@@ -601,6 +601,13 @@ script:
       - text
       required: false
       secret: false
+    - auto: PREDEFINED
+      defaultValue: 'true'
+      description: Whether to download the report immediately after generating it or not. If the report is taking more than 10 seconds to be ready, set as *false*.
```
Comment (kirbles19) [inline Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml:606]: If true, downloads the report immediately after the report is generated. The default is "true". If the report takes longer than 10 seconds to generate, set to "false".

File: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml [hunk 16]
```diff
@@ -656,6 +675,13 @@ script:
       - text
       required: false
       secret: false
+    - auto: PREDEFINED
+      defaultValue: 'true'
+      description: Whether to download the report immediately or not. If the report is taking more than 10 seconds to be ready, set as *false*.
```
Comment (kirbles19) [inline Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml:680]: If true, downloads the report immediately after the report is generated. The default is "true". If the report takes longer than 10 seconds to generate, set to "false".

File: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml [hunk 17]
```diff
@@ -711,6 +749,13 @@ script:
...
```
Comment (kirbles19) [inline Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml:754]: If true, downloads the report immediately after the report is generated. The default is "true". If the report takes longer than 10 seconds to generate, set to "false".

File: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml [hunk 18]
```diff
@@ -932,6 +989,68 @@ script:
...
+      - pdf
+      - rtf
+      - xml
+      - html
+      - text
+      description: The report format, default is pdf.
+      defaultValue: pdf
+    outputs:
+    - contextPath: InfoFile.EntryId
+      description: Entry Id of the report file
```
Comment (kirbles19) [inline Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml:1015]: Entry ID of the report file.

File: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml [hunk 19]
```diff
@@ -932,6 +989,68 @@ script:
...
+      type: string
+    - contextPath: InfoFile.Name
+      description: Name of the report file
```
Comment (kirbles19) [inline Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml:1018]: Name of the report file.

File: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml [hunk 20]
```diff
@@ -932,6 +989,68 @@ script:
...
+      type: string
+    - contextPath: InfoFile.Extension
+      description: File extension of the report file
```
Comment (kirbles19) [inline Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml:1021]: File extension of the report file.

File: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml [hunk 21]
```diff
@@ -932,6 +989,68 @@ script:
...
+      type: string
+    - contextPath: InfoFile.Info
+      description: Info about the report file
```
Comment (kirbles19) [inline Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml:1024]: Information about the report file.

File: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml [hunk 22]
```diff
@@ -932,6 +989,68 @@ script:
...
+      type: string
+    - contextPath: InfoFile.Size
+      description: Size of the report file
```
Comment (kirbles19) [inline Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml:1027]: Size of the report file (in bytes).

Is it in bytes?

File: Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml [hunk 23]
```diff
@@ -932,6 +989,68 @@ script:
...
+      type: number
+    - contextPath: InfoFile.Type
+      description: Type of the report file
```
Comment (kirbles19) [inline Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.yml:1030]: The report file type.

**Review by kirbles19 — APPROVED** [review]
Done.

**Comment by kirbles19** [comment]
@Bargenish Done
